[CIR] Handle FunctionToPointerDecay casts (#153657)

Arghnews opened the pull request:
Add upstream support for handling implicit FunctionToPointerDecay casts in ClangIR, for task #153657, migrate code from clangir repo

Passes `check-clang` test target, have run `git-clang-format` from repo as instructed

Appreciate any feedback, my first LLVM PR so any pointers/improvements/where I'm wrong/conventions I should be aware of etc. please let me know, thanks

Notes:
- Omitted `clang::CIRGen::TBAAAccessInfo` from `emitPointerWithAlignment` as it looks like a large task to include that and not necessary
- Omitted `KnownNonNull_t` param too, although this is much smaller but would have to alter all sites where it's currently used in clangir, in llvm
- Copied `emitFunctionDeclLValue` from clangir
- Have just used the simple example test given in the task + the existing `function-to-pointer-decay.c` test, maybe the CHECKs can be improved or more test cases are needed (if so please specify what, am a CIR noob)
- Left `llvm_unreachable("NYI");` as those rather than a longer string, other instances in file have the same

mmha reviewed (commented):
Thanks for working on this! I mostly have style related comments.

andykaylor reviewed (commented):
@mmha covered everything I would have. I have just one minor difference in suggested changes.

Arghnews commented on clang/test/CIR/CodeGen/function-to-pointer-decay.c:
```diff
@@ -0,0 +1,25 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o - | FileCheck %s
```
If I change the test file to:
```c
  // RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o %t.cir
  // RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -Wno-unused-value -emit-llvm %s -o %t.ll
  // RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -Wno-unused-value -fclangir -emit-llvm %s -o %t-cir.ll

  void f(void);
  void test_call_lvalue_cast() {
    (void (*)())f; // Works
    (*(void (*)(void))f)(); // Works
    (*(void (*)(int))f)(42); // Fails with -fclangir -emit-llvm
  }
```


Then it crashes when running `-fclangir -emit-llvm` with `(*(void (*)(int))f)(42);`, top of the crash:


```
dialect-conversion] Legalizing operation : 'cir.call' (0x6480afdaa150) {
[dialect-conversion]   "cir.call"(%5, %7) <{side_effect = 0 : i32}> : (!cir.ptr<!cir.func<(!cir.int<s, 32>)>>, !cir.int<s, 32>) -> ()
[dialect-conversion]   * Fold {
[dialect-conversion]   } -> FAILURE : unable to fold
[dialect-conversion] 
[dialect-conversion]   * Pattern : 'cir.call -> ()' {
!cir.func<(!cir.int<s, 32>)>
```
> clang: /home/justin/code/llvm-project/mlir/lib/Dialect/LLVMIR/IR/LLVMTypes.cpp:219: static LLVMFunctionType mlir::LLVM::LLVMFunctionType::get(Type, ArrayRef<Type>, bool): Assertion `result && "expected non-null result"' failed.

Have attached the full stack with some debugging turned on - I'm not sure if there's better options, I was just exploring:
[crash.txt](https://github.com/user-attachments/files/21847423/crash.txt)

&nbsp;

Is this a bug that hasn't been implemented yet etc.?
What should be done about it?


The [original clangir test](https://github.com/llvm/clangir/blob/main/clang/test/CIR/CodeGen/function-to-pointer-decay.c) only contains:
```c
// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o - | FileCheck %s

void f(void);

void test_call_lvalue_cast() {
  (*(void (*)(int))f)(42);
}
```

Although if I run this test on clangir with `-fclangir -emit-llvm`, it doesn't crash, and outputs:
```ll
declare void @f()                                                                                                                                                                                                                                                                                                                
; Function Attrs: noinline nounwind optnone                                                                                                                                             
define dso_local void @test_call_lvalue_cast() #0 {                                                                                                                                     
  call void @f()                                                                                                                                                                        
  call void @f(i32 42)                                                                                                                                                                  
  ret void                                                                                                                                                                              
}                                                                                                                                                                                       
```

I'm happy to try and fix it if I can be given some pointers as to how/what/where, if that's within the scope of this PR. Or another etc. Or maybe I messed up something on the way.

On a C level, what is the expected behaviour when you cast a function pointer declared as taking no parameters `void f(void)` to one taking one `int` param and call it? [gcc warns](https://godbolt.org/z/6cxo7nqcf)

Let me know what to do on this

andykaylor replied:
> clang: /home/justin/code/llvm-project/mlir/lib/Dialect/LLVMIR/IR/LLVMTypes.cpp:219: static LLVMFunctionType
> mlir::LLVM::LLVMFunctionType::get(Type, ArrayRef, bool): Assertion `result && "expected non-null result"' failed.

I ran into that recently. I think it's because we weren't providing a converted for the `!cir.void` type. I added one in a PR that I merged yesterday.

https://github.com/llvm/llvm-project/blob/0d9b9d1eef17cd9fa0ea5b1b62469703f9a4dda5/clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp#L2255

Can you rebase your patch and try it again.

Arghnews commented:
> Thanks for working on this! I mostly have style related comments.

I appreciate the review (and it being done so fast!), thank you (to all). I've implemented all the formatting changes that @mmha suggested and @andykaylor 's one instead for the last. I had - wrongly - assumed that `git-clang-format` would change the case style of variables I'd edited for me, so hadn't paid that much attention to their case.

I squashed and force pushed my formatting changes out of habit (for cleaner history in the final commit), but will note that next time I'll wait till the end to do that

Have left the one question regarding the tests open

andykaylor commented:
> On a C level, what is the expected behaviour when you cast a function pointer declared as taking no parameters void f(void) to one taking one int param and call it?

I'm pretty sure that's undefined behavior. In general, the compiler tolerates a lot of things like this because a lot of old C code plays very fast and loose with pointers, and code like that in your example, while technically undefined, is often expected to work (in whatever way it always has).

It appears that clang has decided to be more tolerant of this than GCC, though you can get a warning for it if you explicitly add `-Wcast-function-type`.

bcardosolopes reviewed (approved):
LGTM with the other proposed changes

Arghnews committed:
[CIR] Handle FunctionToPointerDecay casts (#153657)

Add upstream support for handling implicit FunctionToPointerDecay casts
in ClangIR

Arghnews commented:
@andykaylor thank you for the suggestion, have rebased and rerun with latest and it works fine now

Have updated the tests - I wasn't sure how closely the test should watch the cir output (whether too close a match is an issue, or not), so please check. I force pushed anyway so that if this is fine, it can be committed as is

mmha reviewed (approved):
LGTM

andykaylor reviewed (approved):
lgtm
